### PR TITLE
[serviceaccount] Annotations as Values Support

### PR DESCRIPTION
**Pull request by hfuss**
I'd like to be able to include annotations on my operator's ServiceAccount as this is a fairly common need for a lot of apps using ServiceAccounts.

Felt like because ServiceAccounts might be special in this regard, I copied `processer.ProcessObjMeta` and modified it to include an annotations template and value.

Feels a bit hacky and open to feedback as to what the right approach should be.

### CODE REVIEW

**Review by arttor — CHANGES_REQUESTED** [review]
Can you please add sample annotations to one of service accounts in `./test_data` and regenerate chart examples with cat test_data/sample-app.yaml | go run ./cmd/helmify examples/app and cat test_data/k8s-operator-kustomize.output | go run ./cmd/helmify examples/operator and push changes to your branch

**Comment by hfuss** [comment]
Thanks for the steer @arttor, this PR should be good to go now in that it has the test_data and examples update you asked for.

**Comment by arttor** [comment]
@hfuss, I apologize for not bringing up this question earlier, but could you please explain the use case for the `serviceAccount` annotation and how it differs from other objects?

I've tried to do some research but there is not much information:
1. [k8s doc says](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#manually-create-a-long-lived-api-token-for-a-serviceaccount) that you can annotate a Secret to use it as api token for `ServiceAccount`
2. [azure-specific](https://azure.github.io/azure-workload-identity/docs/topics/service-account-labels-and-annotations.html#annotations-1) IAM annotations. AWS also uses annotation to map `serviceAccount` to IAM role.


There may be other annotation use cases beyond `serviceAccount` or IAM. Third-party Kubernetes operators rely on annotations for autoscaling and other tasks.

Unfortunately, Helm does not offer a way to add custom annotations to chart objects, and the chart maintainer may not be aware of which operators the chart users will use and which chart objects they will need to annotate. As a result, it is common to add annotations manually after running helm install because all annotations will be retained during the next helm upgrade.

**Comment by hfuss** [comment]
Hey @arttor no worries, thanks for doing the initial research.

Specifically, my use case is I've written an operator which requires cloud access to either AWS or Azure. As you pointed out, being able to add IAM ServiceAccount annotations for those clouds simplifies how an operator / app needs to authenticate and assume roles.

Another use case for an app or an operator with a UI, might be wanting to create an OAuth client from a ServiceAccount within OpenShift: https://access.redhat.com/documentation/en-us/openshift_container_platform/4.3/html/authentication/using-service-accounts-as-oauth-client

In my use of open-source Helm charts such as External DNS, cert-manager, and External Secrets, it was very common for the chart maintainers to expose their ServiceAccount annotations for the same reason - allowing simpler cloud IAM authentication. Users of  helmify writing cloud-aware operators might often want to offer the same IAM authentication methods for their charts

Your workaround to manually patch ServiceAccount annotations ourselves outside of Helm could introduce a lot of complexity if folks are trying to use Helm in a purely declarative / GitOps driven way such as ArgoCD or Flux.

**Comment by arttor** [comment]
Understood. What do you think about the following approach?
- add a [function option](https://github.com/uber-go/guide/blob/master/style.md#functional-options) to `processor.ProcessObjMeta`. Something like `func WithAnnotations(v helmify.Values) MetaOpt ...`
- in `ProcessObjMeta`: template annotations to `values.yaml` if annotations and option are presented.
- in service account processor: call `ProcessObjMeta` with `WithAnnotations` option.

**Comment by hfuss** [comment]
@arttor great feedback, took a stab at the functional option implementation as you described and ready for review.

File: pkg/processor/meta.go [hunk 1]
```diff
@@ -19,8 +20,51 @@ metadata:
...
+	}
+}
+
 // ProcessObjMeta - returns object apiVersion, kind and metadata as helm template.
-func ProcessObjMeta(appMeta helmify.AppMetadata, obj *unstructured.Unstructured) (string, error) {
+func ProcessObjMeta(appMeta helmify.AppMetadata, obj *unstructured.Unstructured, opts ...MetaOpt) (string, error) {
+	options := &options{
+		values:      nil,
+		annotations: false,
+	}
```
Comment (arttor) [inline pkg/processor/meta.go:63]:
```suggestion
	options := &options{}
```

File: examples/operator/templates/serviceaccount.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: {{ include "operator.fullname" . }}-controller-manager
+  labels:
+  {{- include "operator.labels" . | nindent 4 }}
+  annotations:
+    k8s.acme.org/some-meta-data: ACME Inc.
```
Comment (arttor) [inline examples/operator/templates/serviceaccount.yaml:8]: should `k8s.acme.org/some-meta-data: ACME Inc.` be moved to `values.yaml`?

**Review by arttor — CHANGES_REQUESTED** [review]

**Comment by Starefossen** [comment]
Waiting for this 🙏🏻 I need this in order for service account to be able to use [Google Cloud Workload Identity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity).

```
kubectl annotate serviceaccount KSA_NAME \
    --namespace NAMESPACE \
    iam.gke.io/gcp-service-account=GSA_NAME@GSA_PROJECT.iam.gserviceaccount.com
```

File: examples/operator/values.yaml [hunk 3]
```diff
@@ -44,6 +44,8 @@ controllerManager:
     region: east
     type: user-node
   replicas: 1
+  serviceaccount:
```
Comment (Starefossen) [inline examples/operator/values.yaml:47]: @hfuss where does this come from? The generated serviceaccount.yaml says `{{ toYaml .Values.controllerManager.annotations | nindent 4 }}` and I don't see any changes to my `values.yaml` file.

File: examples/operator/templates/serviceaccount.yaml (hunk 2 above)
Comment (Starefossen) [reply examples/operator/templates/serviceaccount.yaml:8]: @arttor does helmify do that with labels?

File: examples/operator/values.yaml (hunk 3 above)
Comment (Starefossen) [reply examples/operator/values.yaml:47]: @arttor @hfuss I believe this this is the correct template https://github.com/arttor/helmify/commit/e7f15294fbaee735362198d050693340e5dd8a61

Comment (Starefossen) [reply examples/operator/values.yaml:47]: I'll do that @arttor 😄

**Comment by Starefossen** [comment]
Superseded by #115

File: examples/operator/values.yaml (hunk 3 above)
Comment (hfuss) [reply examples/operator/values.yaml:47]: Thanks @Starefossen for picking up my slack and seeing this through! 🚀

**Comment by arttor** [comment]
closed with #115

**Pull request closed without merging**